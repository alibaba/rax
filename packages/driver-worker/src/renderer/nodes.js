import { isValidW3CTag, isValidCustomTags } from './tags';


const CUSTOM_TAG_PREFIX = 'a-';

/**
 * Only node name that in whitelist can be created.
 */
export function getTagName(nodeName) {
  const tagName = String(nodeName).toLowerCase();
  if (isValidW3CTag(tagName)) {
    return tagName;
  } else if (isValidCustomTags(tagName)) {
    return CUSTOM_TAG_PREFIX + tagName;
  } else {
    return 'a-unknown';
  }
}

export function createNode(vnode) {
  let node, tagName;
  if (vnode.nodeType === 3) {
    node = document.createTextNode(vnode.data);
  } else if (vnode.nodeType === 1 && (tagName = getTagName(vnode.nodeName))) {
    node = document.createElement(tagName);
  } else if (vnode.nodeType === 8) {
    node = document.createComment(vnode.data);
  }

  return node;
}
