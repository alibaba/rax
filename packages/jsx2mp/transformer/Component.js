const { resolve, relative } = require('path');
const { transformJSX, writeFiles, isCustomComponent, formatConfing } = require('./Transformer');
const removeExt = require('../utils/removeExt');

/**
 * Create component files
 * @param usingComponents {Object} using Components
 * @param distPath {String} dist Path
 * @param rootContext {String} root Path
 */
const createComponent = function(rootContext, distPath, usingComponents) {
  for (let [key, value] of Object.entries(usingComponents)) {
    if (isCustomComponent(value)) {
      const relativePath = relative(rootContext, value);
      const componentDistPath = removeExt(resolve(distPath, relativePath));
      const transformed = transformJSX(value, 'component');
      createComponent(rootContext, distPath, transformed.config.usingComponents);
      transformed.config = formatConfing(transformed.config, rootContext);
      writeFiles(componentDistPath, transformed, rootContext);
    }
  }
};

module.exports = {
  createComponent
};
