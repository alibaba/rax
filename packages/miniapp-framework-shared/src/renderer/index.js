import { setupAppear } from 'runtime-shared/src/appear';
import { setupTap } from './tap';
import { setupTheme } from './theme';
import { setupReachBottom } from './reachBottom';

export {
  setupAppear,
  setupTap,
  setupTheme,
  setupReachBottom,
};
