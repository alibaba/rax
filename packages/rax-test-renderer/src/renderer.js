import { shared } from 'rax';
import ServerDriver from 'driver-server';
import Serializer from './serializer';

const { Instance, Host } = shared;

// For inject Host init options
Host.driver = ServerDriver;

export default {
  create(element) {
    let container = ServerDriver.createBody();
    let rootInstance = Instance.mount(element, container, {});
    // The field corresponding to node and instance is _r, which you can see in rax/src/vdom/instance
    let renderedComponent = rootInstance.__getRenderedComponent();

    renderedComponent.toJSON = () => {
      return new Serializer(container).toJSON();
    };

    renderedComponent.getInstance = () => {
      return renderedComponent._instance;
    };

    renderedComponent.update = (element) => {
      Instance.mount(element, container, {});
    };

    renderedComponent.unmount = () => {
      let component = Instance.get(container);

      if (!component) {
        return false;
      }

      Instance.remove(container);
      component._internal.unmountComponent();

      return true;
    };

    return renderedComponent;
  }
};
