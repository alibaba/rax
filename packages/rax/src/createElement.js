import Host from './vdom/host';
import Element from './vdom/element';
import flattenChildren from './vdom/flattenChildren';
import { warning, throwError, throwMinifiedWarn } from './error';
import { isString, isArray, NOOP } from './types';
import validateChildKeys from './validateChildKeys';

const RESERVED_PROPS = {
  key: true,
  ref: true,
};

export default function createElement(type, config, children) {
  // Reserved names are extracted
  let props = {};
  let propName;
  let key = null;
  let ref = null;

  if (config != null) {
    ref = config.ref === undefined ? null : config.ref;
    key = config.key === undefined ? null : '' + config.key;

    // Remaining properties are added to a new props object
    for (propName in config) {
      if (!RESERVED_PROPS[propName]) {
        props[propName] = config[propName];
      }
    }
  }

  // Children arguments can be more than one
  const childrenLength = arguments.length - 2;
  if (childrenLength > 0) {
    if (childrenLength === 1 && !isArray(children)) {
      props.children = children;
    } else {
      let childArray = children;
      if (childrenLength > 1) {
        childArray = new Array(childrenLength);
        for (var i = 0; i < childrenLength; i++) {
          childArray[i] = arguments[i + 2];
        }
      }
      props.children = flattenChildren(childArray);
    }
  }

  // Resolve default props
  if (type && type.defaultProps) {
    let defaultProps = type.defaultProps;
    for (propName in defaultProps) {
      if (props[propName] === undefined) {
        props[propName] = defaultProps[propName];
      }
    }
  }

  if (type == null) {
    if (process.env.NODE_ENV !== 'production') {
      throwError(`Invalid element type, expected a string or a class/function component but got "${type}".`);
    } else {
      // A empty component replaced avoid break render in production
      type = NOOP;
      throwMinifiedWarn(0);
    }
  }

  if (process.env.NODE_ENV !== 'production') {
    if (isString(ref) && !Host.owner) {
      warning(
        `Adding a string ref "${ref}" that was not created inside render method, or multiple copies of Rax are used.`
      );
    }

    for (let i = 2; i < arguments.length; i ++) {
      validateChildKeys(arguments[i], type);
    }
  }

  return new Element(
    type,
    key,
    ref,
    props,
    Host.owner
  );
}

