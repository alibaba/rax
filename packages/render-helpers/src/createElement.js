import Host from './host';

/**
* For template include
*/
function processPropsData(props = {}, self = {}) {
  const data = Object.create(props.data || {});
  const { pageInstance } = props;

  Object.keys(props.data || {})
    .forEach((key) => {
      const val = props.data[key];
      if (typeof val === 'string') {
        data[key] = pageInstance && typeof pageInstance[val] === 'function'
          ? pageInstance[val].bind(pageInstance) : val;
      }
    });
  return data;
}

export default function(type, props, ...children) {
  if (typeof type === 'object' && type.__esModule) {
    type = type.default;
  } else if (type === '$template') {
    if (props.is) {
      const data = processPropsData(props, this);
      return props.is.call(props.pageInstance, data);
    } else {
      return null;
    }
  }

  // For reduce code size
  // eg: _c(View, undefined, ['xxx'])
  // vs: _c(View, ['xxx'])
  if (Array.isArray(props)) {
    children = props;
    props = undefined;
  }

  // For array mutation
  for (let prop in props) {
    if (Array.isArray(props[prop])) {
      props[prop] = [...props[prop]];
    }
  }

  return Host.render.createElement(type, props, ...children);
}


