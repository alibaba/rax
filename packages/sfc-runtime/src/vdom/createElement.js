import { processPropsData } from '../utils';
import { get } from '../store';

export default function(type, props, children) {
  let createElement;
  if (this && typeof this.createElement === 'function') {
    createElement = this.createElement;
  } else {
    createElement = get('c');
  }

  if (typeof type === 'object' && type.__esModule === true) {
    type = type.default;
  }

  if (type === '$template') {
    if (props.is) {
      const data = processPropsData(props, this);
      return props.is.call(this.ctx, data);
    } else {
      return null;
    }
  }
  // for reduce code size
  // eg: _c(View, undefined, ['xxx'])
  // vs: _c(View, ['xxx'])
  if (Array.isArray(props)) {
    children = props;
    props = undefined;
  }

  /**
   * for array mutation
   */
  for (let prop in props) {
    if (Array.isArray(props[prop])) {
      props[prop] = [...props[prop]];
    }
  }

  return createElement(type, props, children);
}


