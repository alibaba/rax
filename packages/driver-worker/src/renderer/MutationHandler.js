import { sharedNodeMap } from './NodeMap';
import { createNode } from './nodes';
import { addEvent, removeEvent, setPostMessage } from './events';
import { setAttribute } from './attrs';

const TEXT_CONTENT = 'textContent';
const TEXT_CONTENT_ATTR = TEXT_CONTENT in document ? TEXT_CONTENT : 'nodeValue';

/**
 * Returns "attributes" if it was an attribute mutation.
 * "characterData" if it was a mutation to a CharacterData node.
 * And "childList" if it was a mutation to the tree of nodes.
 */
export default class MutationHandler {
  constructor(sender) {
    setPostMessage(sender);
  }

  apply(data) {
    let mutations = data.mutations;
    for (let i = 0; i < mutations.length; i++) {
      // apply mutation
      let mutation = mutations[i];
      this[mutation.type](mutation);
    }
  }

  childList({ target, removedNodes, addedNodes, nextSibling }) {
    let vnode = target;

    if (vnode && vnode.nodeName === 'BODY') {
      document.body.$$id = vnode.$$id;
    }

    let parent = sharedNodeMap.get(vnode);
    if (removedNodes) {
      for (let i = removedNodes.length; i--;) {
        let node = sharedNodeMap.get(removedNodes[i]);
        sharedNodeMap.delete(node);
        if (parent && node) {
          parent.removeChild(node);
        }
      }
    }

    if (addedNodes) {
      for (let i = 0; i < addedNodes.length; i++) {
        let newNode = sharedNodeMap.get(addedNodes[i]);
        if (!newNode) {
          newNode = createNode(addedNodes[i]);
        }

        if (parent) {
          parent.insertBefore(newNode, nextSibling && sharedNodeMap.get(nextSibling) || null);
        }
      }
    }
  }

  attributes({ target, attributeName, newValue, style }) {
    let node = sharedNodeMap.get(target);
    // Node maybe null when node is removed and there is a setInterval change the node that will cause error
    if (!node) return;

    if (style) {
      setAttribute(node, 'style', style);
    }
    if (attributeName) {
      setAttribute(node, attributeName, newValue);
    }
  }

  characterData({ target, newValue }) {
    let node = sharedNodeMap.get(target);
    node[TEXT_CONTENT_ATTR] = newValue;
  }

  addEvent({ target, eventName }) {
    let node = sharedNodeMap.get(target);
    if (!node) return;

    addEvent(node, eventName);
  }

  removeEvent({ target, eventName }) {
    let node = sharedNodeMap.get(target);
    if (!node) return;

    removeEvent(node, eventName);
  }

  canvasRenderingContext2D({ target, method, args, properties }) {
    let canvas = sharedNodeMap.get(target);
    if (!canvas) return;

    let context = canvas.getContext('2d');

    if (properties) {
      for (let key in properties) {
        if (properties.hasOwnProperty(key)) {
          context[key] = properties[key];
        }
      }
    }

    if (method) {
      context[method].apply(context, args);
    }
  }

  lottieRenderingContext({ target, method, args }) {
    let lottie = sharedNodeMap.get(target);

    if (method) {
      lottie[method].apply(lottie, args);
    }
  }
}
