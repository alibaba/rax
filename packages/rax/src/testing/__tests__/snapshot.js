import createElement from '../../createElement';
import Component from '../../vdom/component';
import renderer from '../renderer';

const STATUS = {
  NORMAL: 'normal',
  HOVERED: 'hovered',
};

class Link extends Component {
  constructor() {
    super();

    this._onMouseEnter = this._onMouseEnter.bind(this);
    this._onMouseLeave = this._onMouseLeave.bind(this);

    this.state = {
      class: STATUS.NORMAL,
    };
  }

  _onMouseEnter() {
    this.setState({class: STATUS.HOVERED});
  }

  _onMouseLeave() {
    this.setState({class: STATUS.NORMAL});
  }

  render() {
    return (
      <a
        className={this.state.class}
        href={this.props.page || '#'}
        onMouseEnter={this._onMouseEnter}
        onMouseLeave={this._onMouseLeave}>
        {this.props.children}
      </a>
    );
  }
}

describe('snapshot', () => {
  beforeEach(function() {
    jest.useFakeTimers();
  });

  afterEach(function() {
    jest.useRealTimers();
  });

  test('Link changes the class when hovered', () => {
    const component = renderer.create(
      <Link page="https://example.com">Example</Link>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.eventListeners.mouseenter();
    jest.runAllTimers();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.eventListeners.mouseleave();
    jest.runAllTimers();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

