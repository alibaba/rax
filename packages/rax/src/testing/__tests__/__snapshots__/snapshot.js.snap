// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Link changes the class when hovered 1`] = `
Object {
  "attributes": Object {
    "class": "normal",
    "href": "https://example.com",
  },
  "children": Array [
    "Example",
  ],
  "eventListeners": Object {
    "mouseenter": [Function],
    "mouseleave": [Function],
  },
  "tagName": "A",
}
`;

exports[`snapshot Link changes the class when hovered 2`] = `
Object {
  "attributes": Object {
    "class": "hovered",
    "href": "https://example.com",
  },
  "children": Array [
    "Example",
  ],
  "eventListeners": Object {
    "mouseenter": [Function],
    "mouseleave": [Function],
  },
  "tagName": "A",
}
`;

exports[`snapshot Link changes the class when hovered 3`] = `
Object {
  "attributes": Object {
    "class": "normal",
    "href": "https://example.com",
  },
  "children": Array [
    "Example",
  ],
  "eventListeners": Object {
    "mouseenter": [Function],
    "mouseleave": [Function],
  },
  "tagName": "A",
}
`;
